docs/developers/napari_architecture.md

```diff
@@ -0,0 +1,164 @@
+        )
+
+    # for each attribute, we create a `@property` getter/setter
+    # so that we can emit the appropriate event when that attribute
+    # is changed using the syntax: ``weather.attribute = new_value``
+    @property
+    def temperature(self):
+        """Degrees in Fahrenheit."""
+        return self._temperature
+
+    @temperature.setter
+    def temperature(self, value):
+        self._temperature = value
+        # emit the temperature "changed" event
+        self.events.temperature(value=value)
+
+    @property
+    def humidity(self):
+        """Percent humidity."""
+        return self._humidity
+
+    @humidity.setter
+    def humidity(self, value):
+        self._humidity = value
+        # emit the humidity "changed" event
+        self.events.humidity(value=value)
+
+    @property
+    def wind(self):
+        """Wind speed in mph."""
+        return self._wind
+
+    @wind.setter
+    def wind(self, value):
+        self._wind = value
+        # emit the wind "changed" event
+        self.events.wind(value=value)
+```
+
+Another object can then "listen" for changes in our weather model and register
+a callback function with the event emitter of the attribute they would like
+to watch:
+
+```python
+# create an instance of the model
+weather = Weather(temperature=72, humidity=65, wind=0)
+
+# define some callback that should respond to changes in the model
+def hurricane_watch(event):
+    if event.value > 74:
+        print("Hurricane! Evacuate!")
+
+# register our callback with the model
+weather.events.wind.connect(hurricane_watch)
+
+# now, everytime weather.wind is changed, hurricane_watch is called
+weather.wind = 90  # prints: "Hurricane! Evacuate!"
+```
+
+This method is very customizable but requires a lot of boilerplate. The
+generic base model `EventedModel` was added to reduce this and
+"standardize" this change/emit pattern. The `EventedModel` provides the
+following features:
+
+* type validation and coersion on class instantiation and attribute assignment
+* event emission after successful attribute assignment
+
+Using `EventedModel` would reduce the above `weather` class code to:
+
+```python
+class weather(EventedModel):
+    """A simple model to track changes in the weather.
+
+    Parameters
+    ----------
+    temperature : float
+        Degrees in Fahrenheit.
+    humidity : float
+        Percent humidity
+    wind : float
+    """
+    temperature: float
+    humidity: float
+    wind: float
+```
+
+Currently most of the models in `napari/components/` are `EventedModels` but
+not the layer models although there is intention to convert these to
+`EventedModels` in the future.
+
+## Qt models
+
+Qt classes are responsible for all napari's user interface elements. There is
+generally one to one mapping between Python models and qt models in napari, for
+example Python model `Dims` and Qt model `QtDims`.
+The Qt class can register callbacks such that when an attribute of the
```

Comment:
If I understand correctly, most of the Qt classes (at least those that implement GUI components), are intended to be the *view* (and often controller) part of the application model, which is roughly model-view-controller.

You effectively mention this at the top of the document, so explicitly tying this section back to that summary seems like a good idea.

Reply:
Adding an example of `the appropriate actions` here would be great. E.g. `dims.current_step = (5, 0, 0)` will cause the `QtDims` object to update the position of the slider representing the first dimesion.

Reply:
To clarify, does 'application model' mean 'model viewer and controller' ?

Reply:
Oops, I shouldn't have said application *model* - that's just confusing - but yes you're right. Maybe *application architecture* would be better as MVC is an architectural pattern.